File: app/collector.py
"""
Domain Collector - Handles domain collection from Gmail inbox
"""

import re
import asyncio
import logging
from typing import Dict, List, Optional, Callable, Tuple
from collections import defaultdict

from googleapiclient.errors import HttpError

from app.models import ThreadMetadata, DomainInfo, CollectionConfig


logger = logging.getLogger(__name__)


class DomainCollector:
    """Handles domain collection from Gmail inbox"""

    def __init__(
        self,
        service,  # Gmail API service object
        config: CollectionConfig,
        progress_callback: Optional[Callable] = None
    ):
        self.service = service
        self.config = config
        self.progress_callback = progress_callback

        # Results - exposed for GmailService to access
        self.threads_by_id: Dict[str, ThreadMetadata] = {}
        self.threads_by_domain: Dict[str, List[str]] = defaultdict(list)
        self.interrupted = False

    # === Main Entry Point ===

    async def collect(self) -> Dict[str, DomainInfo]:
        """Orchestrates the collection process"""
        total_thread_count = await self._get_total_thread_count()
        effective_total = min(self.config.limit, total_thread_count) if self.config.limit else total_thread_count

        message = f"Starting domain collection (limit: {self.config.limit} threads)..." if self.config.limit else "Starting domain collection..."
        await self._report_progress("collection_started", {
            "message": message,
            "total_threads": effective_total,
            "limit": self.config.limit
        })

        # Clear any previous state
        self.threads_by_id.clear()
        self.threads_by_domain.clear()

        domain_data = defaultdict(lambda: {'count': 0, 'subjects': []})
        page_token = None
        total_threads = 0

        while not self.interrupted:
            try:
                threads, next_page_token = await self._fetch_thread_page(page_token)

                if not threads:
                    break

                for thread in threads:
                    if self.interrupted:
                        break

                    thread_id = thread['id']
                    metadata = await self._get_thread_metadata(thread_id)

                    if metadata is None:
                        continue

                    if not self._should_include(metadata):
                        continue

                    # Store thread
                    self._store_thread(metadata)

                    # Update domain data
                    domain_data[metadata.domain]['count'] += 1
                    if metadata.subject not in domain_data[metadata.domain]['subjects'] and len(domain_data[metadata.domain]['subjects']) < 3:
                        domain_data[metadata.domain]['subjects'].append(metadata.subject)

                    total_threads += 1

                    # Send progress update
                    await self._report_progress("thread_processed", {
                        "thread_id": thread_id,
                        "domain": metadata.domain,
                        "subject": metadata.subject[:60] + "..." if len(metadata.subject) > 60 else metadata.subject,
                        "processed_threads": total_threads,
                        "total_threads": effective_total,
                        "unique_domains": len(domain_data)
                    })

                    # Check limit
                    if self.config.limit and total_threads >= self.config.limit:
                        logger.info(f"Reached limit of {self.config.limit} threads, stopping collection")
                        break

                    # Yield control every 10 threads
                    if total_threads % 10 == 0:
                        await asyncio.sleep(0)

                # Check if we hit limit
                if self.config.limit and total_threads >= self.config.limit:
                    break

                page_token = next_page_token
                if not page_token:
                    break

            except HttpError as error:
                await self._report_progress("error", {"message": f"Error fetching threads: {error}"})
                break

        # Build results
        result = self._build_results(domain_data)

        limit_msg = f" (limited to {self.config.limit})" if self.config.limit else ""
        await self._report_progress("collection_completed", {
            "processed_threads": total_threads,
            "total_threads": effective_total,
            "unique_domains": len(result),
            "message": f"Collection complete{limit_msg}: {total_threads:,} threads processed, {len(result):,} unique domains"
        })

        logger.debug(f"Stored {len(self.threads_by_id)} threads across {len(self.threads_by_domain)} domains in memory")

        return result

    # === Thread Fetching ===

    async def _get_total_thread_count(self) -> int:
        """Get inbox thread count for progress reporting"""
        try:
            inbox_info = await asyncio.to_thread(
                lambda: self.service.users().labels().get(
                    userId='me',
                    id='INBOX'
                ).execute()
            )
            return inbox_info.get('threadsTotal', 0)
        except Exception as e:
            logger.warning(f"Could not get inbox thread count: {e}")
            return 0

    async def _fetch_thread_page(self, page_token: Optional[str]) -> Tuple[List[Dict], Optional[str]]:
        """Fetch a page of threads, returns (threads, next_page_token)"""
        results = await asyncio.to_thread(
            lambda: self.service.users().threads().list(
                userId='me',
                maxResults=100,
                pageToken=page_token,
                q='in:inbox'
            ).execute()
        )

        threads = results.get('threads', [])
        next_page_token = results.get('nextPageToken')

        return threads, next_page_token

    async def _get_thread_metadata(self, thread_id: str) -> Optional[ThreadMetadata]:
        """Fetch and parse metadata for a single thread"""
        thread_data = await asyncio.to_thread(
            lambda: self.service.users().threads().get(
                userId='me',
                id=thread_id,
                format='metadata',
                metadataHeaders=['From', 'Subject']
            ).execute()
        )

        messages = thread_data.get('messages', [])
        if not messages:
            return None

        first_message = messages[0]
        headers = {h['name']: h['value'] for h in first_message['payload'].get('headers', [])}
        sender = headers.get('From', '(Unknown Sender)')
        subject = headers.get('Subject', '(No Subject)')
        sender_email = self.extract_email_address(sender)

        # Get labels from both thread and message level
        thread_label_ids = thread_data.get('labelIds', [])
        first_message_label_ids = first_message.get('labelIds', [])
        all_label_ids = list(set(thread_label_ids + first_message_label_ids))

        domain = self.extract_domain(sender_email)
        if not domain:
            return None

        return ThreadMetadata(
            thread_id=thread_id,
            domain=domain,
            subject=subject,
            sender=sender_email,
            message_count=len(messages),
            label_ids=all_label_ids
        )

    # === Filtering ===

    def _is_protected(self, label_ids: List[str]) -> bool:
        """Check if thread is protected by labels (IMPORTANT, STARRED, custom)"""
        # Always protect IMPORTANT and STARRED
        if 'IMPORTANT' in label_ids or 'STARRED' in label_ids:
            logger.debug("Protected by IMPORTANT/STARRED")
            return True

        # Skip label protection if disabled
        if not self.config.use_label_protection:
            return False

        # Check for custom user labels
        custom_labels = [label for label in label_ids if label.startswith('Label_')]
        if not custom_labels:
            return False

        # If specific labels provided, only protect those
        if self.config.protected_label_ids is not None:
            protected = any(label in self.config.protected_label_ids for label in custom_labels)
            if protected:
                matching = [l for l in custom_labels if l in self.config.protected_label_ids]
                logger.debug(f"Protected by selected labels: {matching}")
            return protected

        # Otherwise protect any custom label (default behavior)
        logger.debug(f"Protected by custom labels: {custom_labels}")
        return True

    def _is_excluded(self, domain: str) -> bool:
        """Check if domain is in exclusion list"""
        return domain in self.config.excluded_domains

    def _should_include(self, metadata: ThreadMetadata) -> bool:
        """Determine if thread should be included in results"""
        if self._is_protected(metadata.label_ids):
            logger.debug(f"COLLECT - SKIPPING PROTECTED thread {metadata.thread_id} with labels: {metadata.label_ids}")
            return False

        if self._is_excluded(metadata.domain):
            logger.debug(f"COLLECT - SKIPPING EXCLUDED domain: {metadata.domain}")
            return False

        return True

    # === Storage ===

    def _store_thread(self, metadata: ThreadMetadata) -> None:
        """Store thread in internal dictionaries"""
        self.threads_by_id[metadata.thread_id] = metadata
        self.threads_by_domain[metadata.domain].append(metadata.thread_id)

    # === Results ===

    def _build_results(self, domain_data: Dict) -> Dict[str, DomainInfo]:
        """Aggregate stored threads into DomainInfo objects"""
        result = {}

        for domain, data in domain_data.items():
            thread_ids = self.threads_by_domain.get(domain, [])
            threads = []

            for thread_id in thread_ids:
                metadata = self.threads_by_id.get(thread_id)
                if metadata:
                    threads.append({
                        'thread_id': thread_id,
                        'subject': metadata.subject,
                        'sender': metadata.sender,
                        'message_count': metadata.message_count
                    })

            result[domain] = DomainInfo(
                domain=domain,
                count=data['count'],
                threads=threads
            )

        return result

    # === Progress ===

    async def _report_progress(self, event: str, data: Dict) -> None:
        """Send progress update if callback is set"""
        if self.progress_callback:
            await self.progress_callback(event, data)

    # === Utilities ===

    @staticmethod
    def extract_email_address(sender: str) -> str:
        """Extract email from 'Name <email@domain.com>' format"""
        match = re.search(r'<([^>]+)>', sender)
        if match:
            return match.group(1).lower()
        return sender.strip().lower()

    @staticmethod
    def extract_domain(email: str) -> str:
        """Extract domain from email address"""
        if '@' in email:
            return email.split('@')[1].lower()
        return ''

File: src/lib/gmail/collector.js
/**
 * Domain Collector - Handles domain collection from Gmail inbox
 */

import { getInboxInfo, listThreads, getThread } from './api.js';
import { ThreadsList, Thread, CleanupThread, DomainResult, CollectionResult } from '../models/index.js';
import {
  SUBJECT_TRUNCATE_COLLECTOR,
  MILESTONE_LOG_INTERVAL,
  THREAD_PAGE_SIZE,
} from '../constants.js';
import { getErrorMessage } from '../errors.js';

export class DomainCollector {
  constructor(config, progressCallback = null) {
    this.config = config;
    this.progressCallback = progressCallback;

    // Results - exposed for cleanup to access
    this.threadsById = {};      // threadId -> Thread
    this.threadsByDomain = {};  // domain -> [threadId, ...]
    this.interrupted = false;

    // Pollable progress state — UI reads this via setInterval
    this.progress = {
      processedThreads: 0,
      totalThreads: 0,
      uniqueDomains: 0,
      currentDomain: '',
      status: 'idle', // 'idle' | 'running' | 'completed' | 'error'
      errorMessage: null,
    };
  }

  // === Main Entry Point ===

  async collect() {
    const totalThreadCount = await this._getTotalThreadCount();
    const effectiveTotal = this.config.limit
      ? Math.min(this.config.limit, totalThreadCount)
      : totalThreadCount;

    const message = this.config.limit
      ? `Starting domain collection (limit: ${this.config.limit} threads)...`
      : 'Starting domain collection...';

    await this._reportProgress('collection_started', {
      message,
      total_threads: effectiveTotal,
      limit: this.config.limit,
    });

    // Clear any previous state
    this.threadsById = {};
    this.threadsByDomain = {};

    // Initialize pollable progress
    this.progress.status = 'running';
    this.progress.totalThreads = effectiveTotal;
    this.progress.processedThreads = 0;
    this.progress.uniqueDomains = 0;
    this.progress.currentDomain = '';
    this.progress.errorMessage = null;

    const domainCounts = {}; // domain -> count
    let pageToken = null;
    let totalThreads = 0;

    while (!this.interrupted) {
      try {
        const page = await this._fetchThreadPage(pageToken);

        if (page.threadIds.length === 0) {
          break;
        }

        for (const threadId of page.threadIds) {
          if (this.interrupted) break;

          const thread = await this._getThread(threadId);

          if (thread === null) continue;
          if (!this._shouldInclude(thread)) continue;

          // Store thread
          this._storeThread(thread);

          // Track domain count
          const domain = thread.getDomain();
          domainCounts[domain] = (domainCounts[domain] || 0) + 1;

          totalThreads += 1;

          // Update pollable progress in-place (no await, no callback)
          this.progress.processedThreads = totalThreads;
          this.progress.uniqueDomains = Object.keys(domainCounts).length;
          this.progress.currentDomain = domain;

          // Milestone logging — infrequent, so callback cost is fine
          if (totalThreads % MILESTONE_LOG_INTERVAL === 0) {
            await this._reportProgress('milestone', {
              processed_threads: totalThreads,
              total_threads: effectiveTotal,
              unique_domains: Object.keys(domainCounts).length,
            });
          }

          // Check limit
          if (this.config.limit && totalThreads >= this.config.limit) {
            break;
          }
        }

        // Check if we hit limit
        if (this.config.limit && totalThreads >= this.config.limit) {
          break;
        }

        pageToken = page.nextPageToken;
        if (!pageToken) break;
      } catch (error) {
        this.progress.status = 'error';
        this.progress.errorMessage = getErrorMessage(error);
        await this._reportProgress('error', {
          message: `Error fetching threads: ${getErrorMessage(error)}`,
        });
        break;
      }
    }

    // Build results
    const result = this._buildResults(domainCounts);

    const limitMsg = this.config.limit ? ` (limited to ${this.config.limit})` : '';
    this.progress.status = 'completed';

    await this._reportProgress('collection_completed', {
      processed_threads: totalThreads,
      total_threads: effectiveTotal,
      unique_domains: Object.keys(result.domainResults).length,
      message: `Collection complete${limitMsg}: ${totalThreads.toLocaleString()} threads processed, ${Object.keys(result.domainResults).length.toLocaleString()} unique domains`,
    });

    return result;
  }

  // === Thread Fetching ===

  async _getTotalThreadCount() {
    try {
      const inboxInfo = await getInboxInfo();
      return inboxInfo.threadsTotal || 0;
    } catch (e) {
      console.warn('Could not get inbox thread count:', e);
      return 0;
    }
  }

  async _fetchThreadPage(pageToken) {
    const raw = await listThreads({
      maxResults: THREAD_PAGE_SIZE,
      pageToken,
      q: 'in:inbox',
    });

    return new ThreadsList(raw);
  }

  async _getThread(threadId) {
    const raw = await getThread(threadId, {
      format: 'metadata',
      metadataHeaders: ['From', 'Subject'],
    });

    const thread = new Thread(threadId, raw);
    if (thread.isEmpty() || !thread.getDomain()) return null;

    return thread;
  }

  // === Filtering ===

  _isProtected(labelIds) {
    // Always protect IMPORTANT and STARRED
    if (labelIds.includes('IMPORTANT') || labelIds.includes('STARRED')) {
      return true;
    }

    // Skip label protection if disabled
    if (!this.config.useLabelProtection) {
      return false;
    }

    // Check for custom user labels
    const customLabels = labelIds.filter((l) => l.startsWith('Label_'));
    if (customLabels.length === 0) return false;

    // If specific labels provided, only protect those
    if (this.config.protectedLabelIds !== null) {
      return customLabels.some((l) => this.config.protectedLabelIds.has(l));
    }

    // Otherwise protect any custom label (default behavior)
    return true;
  }

  _isExcluded(domain) {
    return this.config.excludedDomains.has(domain);
  }

  _shouldInclude(thread) {
    if (this._isProtected(thread.getLabelIds())) return false;
    if (this._isExcluded(thread.getDomain())) return false;
    return true;
  }

  // === Storage ===

  _storeThread(thread) {
    const domain = thread.getDomain();
    this.threadsById[thread.threadId] = thread;
    if (!this.threadsByDomain[domain]) {
      this.threadsByDomain[domain] = [];
    }
    this.threadsByDomain[domain].push(thread.threadId);
  }

  // === Results ===

  _buildResults(domainCounts) {
    const domainResults = {};

    for (const [domain, count] of Object.entries(domainCounts)) {
      const threadIds = this.threadsByDomain[domain] || [];
      const threads = threadIds
        .map((id) => this.threadsById[id])
        .filter(Boolean)
        .map((thread) => CleanupThread.fromThread(thread));

      domainResults[domain] = new DomainResult({ domain, count, threads });
    }

    return new CollectionResult(domainResults, this.threadsById, this.threadsByDomain);
  }

  // === Progress ===

  async _reportProgress(event, data) {
    if (this.progressCallback) {
      await this.progressCallback(event, data);
    }
  }

}
